client/src/client.rs

```diff
@@ -612,6 +612,23 @@ impl Iterator for EventIterator {
     }
 }
 
+impl Drop for EventIterator {
+    fn drop(&mut self) {
+        let mut close = || -> eyre::Result<()> {
+            self.stream.close(None)?;
+            let mes = self.stream.read_message()?;
```

Comment:
What I'm more concerned is that we should have some timeout if we're connecting to malicious peer that will never terminate the connection. At that point we should force closing the connection from the client side

Reply:
We should make sure the peer has timeout for terminating websocket connections as well

Reply:
> Also tungstenite WebSocket::close documentation seems to me to indicate that it might be better to depend on Error::ConnectionClosed

According to tungstentite docs do you suggest something like this?

```rust
self.stream.close(None)?;
let mes = self.stream.read_message()?;
if !mes.is_close() {
    return Err(eyre!(
        "Server hasn't sent `Close` message for websocket handshake"
    ));
}
if !matches!(self.stream.read_message(), Err(tungstentite::Error::ConnectoinClosed)) {
    return Err(eyre!("Connection should be closed"));
}
```

We had to read the message twice. Don't know if it's worth it

Reply:
> We should make sure the peer has timeout for terminating websocket connections as well

That's interesting. Right now client sends `Close` message to stop event translation. How (and why) should peer decide to drop connection on it's own?

Reply:
> What I'm more concerned is that we should have some timeout if we're connecting to malicious peer that will never terminate the connection. At that point we should force closing the connection from the client side

After reading some docs and responses from tungstentite authors I believe, that the best way to do it -- is to use async stream. And that is the whole issue #1256 . I'll mention your idea in the comments.

Without that the only possible solution I can see is to use threads. I think, that will be just a code to delete in future, when client will be moved to async. So maybe just let it be implemented in #1256 ?

Reply:
> That's interesting. Right now client sends Close message to stop event translation. How (and why) should peer decide to drop connection on it's own?

We expect there to be Byzantine faults, which includes clients and peers. One method of attack is sending a message for an unreasonably long time, so we want to have a time-out (just like with Sumeragi), when we just assume the message isn't coming at all. 

> After reading some docs and responses from tungstentite authors I believe, that the best way to do it -- is to use async stream. And that is the whole issue https://github.com/hyperledger/iroha/issues/1256 . I'll mention your idea in the comments.

Sounds good. 

> Without that the only possible solution I can see is to use threads. I think, that will be just a code to be deleted in future, when client will be moved to async. So maybe just let it be implemented in https://github.com/hyperledger/iroha/issues/1256 ?

Issue #1256 is relatively high priority for Blockchain explorer.